File: packages/lib-angular/projects/avatars/src/lib/avatars/avatar-ring.ts
import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";

import { Component, Input } from "@angular/core";

import type { AvatarProps } from "../avatar.utils";

import { defaultAvatarProps } from "../avatar.utils";
import { generateColors, SIZE } from "./avatar-ring.utils";

@Component({
  selector: "avatar-ring, AvatarRing",
  template: `
    <svg
      fill="none"
      role="img"
      xmlns="http://www.w3.org/2000/svg"
      [attr.viewBox]='"0 0 " + SIZE + " " + SIZE'
      [attr.width]="size"
      [attr.height]="size"
    >
      <ng-container *ngIf="title">
        <title>{{name}}</title>
      </ng-container>

      <mask
        id="mask__ring"
        maskUnits="userSpaceOnUse"
        [attr.x]="0"
        [attr.y]="0"
        [attr.width]="SIZE"
        [attr.height]="SIZE"
      >
        <rect
          fill="#FFFFFF"
          [attr.width]="SIZE"
          [attr.height]="SIZE"
          [attr.rx]="square ? undefined : SIZE * 2"
        ></rect>
      </mask>

      <g mask="url(#mask__ring)">
        <path
          d="M0 0h90v45H0z"
          [attr.fill]="ringColors(name, colors)[0]"
        ></path>

        <path
          d="M0 45h90v45H0z"
          [attr.fill]="ringColors(name, colors)[1]"
        ></path>

        <path
          d="M83 45a38 38 0 00-76 0h76z"
          [attr.fill]="ringColors(name, colors)[2]"
        ></path>

        <path
          d="M83 45a38 38 0 01-76 0h76z"
          [attr.fill]="ringColors(name, colors)[3]"
        ></path>

        <path
          d="M77 45a32 32 0 10-64 0h64z"
          [attr.fill]="ringColors(name, colors)[4]"
        ></path>

        <path
          d="M77 45a32 32 0 11-64 0h64z"
          [attr.fill]="ringColors(name, colors)[5]"
        ></path>

        <path
          d="M71 45a26 26 0 00-52 0h52z"
          [attr.fill]="ringColors(name, colors)[6]"
        ></path>

        <path
          d="M71 45a26 26 0 01-52 0h52z"
          [attr.fill]="ringColors(name, colors)[7]"
        ></path>

        <circle
          [attr.cx]="45"
          [attr.cy]="45"
          [attr.r]="23"
          [attr.fill]="ringColors(name, colors)[8]"
        ></circle>
      </g>
    </svg>
  `,
})
export class AvatarRing {
  SIZE = SIZE;

  @Input() size: Omit<AvatarProps, 'variant'>['size'] = defaultAvatarProps['size'];
  @Input() title: Omit<AvatarProps, 'variant'>['title'] = defaultAvatarProps['title'];
  @Input() name: Omit<AvatarProps, 'variant'>['name'] = defaultAvatarProps['name'];
  @Input() square: Omit<AvatarProps, 'variant'>['square'] = defaultAvatarProps['square'];
  @Input() colors: Omit<AvatarProps, 'variant'>['colors'] = defaultAvatarProps['colors'];

  ringColors(name, colors) {
    return generateColors(colors, name);
  }
}

@NgModule({
  declarations: [AvatarRing],
  imports: [CommonModule],
  exports: [AvatarRing],
})
export class AvatarRingModule {}

File: packages/lib-leptos/src/avatar_ring.rs
use crate::avatar_ring_utils::*;
use crate::avatar_utils::*;
use leptos::*;

#[component]
pub fn AvatarRing(
    cx: Scope,
    name: String,
    colors: Vec<&'static str>,
    square: bool,
) -> impl IntoView {
    let num_from_name = hash_code(&name);
    let generated_colors = generate_colors(&name, colors);

    let outer_radius = if square == true { 0.0 } else { (SIZE as f64) * 2.0 };

    view! { cx,
      <svg
        fill="none"
        role="img"
        xmlns="http://www.w3.org/2000/svg"
        viewBox=format!("0 0 {} {}",  SIZE, SIZE)
        width=200
        height=200
      >
        <mask
        id=format!("mask__ring_{}", num_from_name)
          maskUnits="userSpaceOnUse"
          x=0
          y=0
          width=SIZE
          height=SIZE
        >
          <rect
            fill="#FFFFFF"
            width=SIZE
            height=SIZE
            rx=outer_radius
          />
        </mask>

        <g mask=format!("url(#mask__ring_{})", num_from_name)>

          <path
            d="M0 0h90v45H0z"
            fill=generated_colors[0].clone()
          />

          <path
            d="M0 45h90v45H0z"
            fill=generated_colors[1].clone()
          />

          <path
            d="M83 45a38 38 0 00-76 0h76z"
            fill=generated_colors[2].clone()
          />

          <path
            d="M83 45a38 38 0 01-76 0h76z"
            fill=generated_colors[3].clone()
          />

          <path
            d="M77 45a32 32 0 10-64 0h64z"
            fill=generated_colors[4].clone()
          />

          <path
            d="M77 45a32 32 0 11-64 0h64z"
            fill=generated_colors[5].clone()
          />

          <path
            d="M71 45a26 26 0 00-52 0h52z"
            fill=generated_colors[6].clone()
          />

          <path
            d="M71 45a26 26 0 01-52 0h52z"
            fill=generated_colors[7].clone()
          />

          <circle
            cx=45
            cy=45
            r=23
            fill=generated_colors[8].clone()
          />
        </g>
      </svg>
    }
}
